src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,184 @@
+        return findNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        if (isEmpty()) {
+            setFirstValue(value);
+            return value;
+        }
+        T removedValue = remove(index);
+        add(value, index);
+        return removedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> node = findNodeByIndex(index);
+        T value = node.item;
+        unlink(node);
+        return value;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        try {
+            unlink(findNodeByValue(object));
+        } catch (RuntimeException e) {
+            return false;
+        }
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private void checkAddIndex(int index) {
+        if (index > size || index < 0) {
+            throw new IndexOutOfBoundsException("Your index was " + index
+                    + ", but size is " + size);
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException("Your index was " + index
+                    + ", but size is " + size);
+        }
+    }
+
+    private Node<T> findNodeByValue(T value) {
+        Node<T> node = first;
+        for (int i = 0; i < size; i++) {
+            if (value == node.item || value != null
+                    && value.equals(node.item)) {
+                return node;
+            }
+            node = node.next;
+        }
+        throw new RuntimeException("Element didn't find");
+    }
+
+    private Node<T> findNodeByIndex(int index) {
+        int part = size % 2 == 0 ? size / 2 : size / 2 + 1;
+        if (index < part) {
+            Node<T> node = first;
+            for (int i = 0; i < index; i++) {
+                node = node.next;
+            }
+            return node;
+        }
+        Node<T> node = last;
+        for (int i = 0; i < size - index - 1; i++) {
+            node = node.prev;
+        }
+        return node;
+    }
+
+    private void linkBefore(T element, Node<T> next) {
+        Node<T> prev = next.prev;
+        Node<T> newNode = new Node<>(prev, element, next);
+        next.prev = newNode;
+        if (prev == null) {
+            first = newNode;
+            size++;
+            return;
+        }
+        prev.next = newNode;
+        size++;
+    }
```

Comment:
you use this method only once, do you really need a separate method for this code?

Reply:
in this case the method will be too large